Added the functionality to exclude files from backup

SailReal opened the pull request:
Execute the command with excluding files like this:
```
./conserve backup archiv/ folder/ "f*"
./conserve backup archiv/ folder/ foo,bar
```

With the option verbose (-v) the output will show something like this:
```
Backup folder
Skipping foo
Backup folder/barBaz
```

sourcefrog commented on README.md:
```diff
@@ -47,6 +47,8 @@ source tree), the backup is considered *complete*.
 
     conserve backup /backup/home.cons ~
 
+Optionally provide as third parameter a comma separated list of [glob patterns](https://en.wikipedia.org/wiki/Glob_(programming) to exclude files from backup.
```
I think I'd rather a `--exclude` option rather than a third parameter, as it can more easily apply to other commands, and is more consistent with eg rsync and duplicity.

sourcefrog replied:
Also please wrap the text in this file.

sourcefrog commented on src/backup.rs:
```diff
@@ -11,10 +11,13 @@ use super::*;
 use index;
 use sources;
 
+extern crate globset;
```
The crate should be included into `lib.rs`

sourcefrog commented on src/backup.rs:
```diff
@@ -11,10 +11,13 @@ use super::*;
 
+extern crate globset;
+
+use self::globset::{Glob, GlobSet, GlobSetBuilder};
```
Should be just `use globset::...`, it's not conceptually under `self`.

sourcefrog commented on src/backup.rs:
```diff
@@ -26,22 +29,52 @@ struct Backup {
 
 impl BackupOptions {
-    pub fn backup(&self, archive_path: &Path, source: &Path, report: &Report) -> Result<()> {
+    pub fn backup(&self, archive_path: &Path, source: &Path, report: &Report, excludes: Option<Vec<&str>>) -> Result<()> {
```
The `BackupOptions` is using the options builder pattern, which is a way to avoid having an over-long parameter set for `backup()`. Please set the excludes into the options into a field in `BackupOptions` from a separate method before running `backup`.

sourcefrog commented on src/backup.rs:
```diff
@@ -26,22 +29,52 @@ struct Backup {
+            Some(excludes) => {
+                let mut builder = GlobSetBuilder::new();
+                for exclude in excludes {
+                    builder.add(Glob::new(exclude).chain_err(|| "Failed to parse exclude value")?);
```
Include the value that didn't parse.

sourcefrog commented on src/backup.rs:
```diff
@@ -26,22 +29,52 @@ struct Backup {
     }
+
+
+    fn get_excludes(excludes: Option<Vec<&str>>) -> Result<Option<GlobSet>> {
```
This isn't backup-specific and should be somewhere else, maybe in a new `excludes.rs`. Also, it can then have unit tests there.

sourcefrog commented on src/backup.rs:
```diff
@@ -26,22 +29,52 @@ struct Backup {
         let source_iter = try!(sources::iter(source, report));
         for entry in source_iter {
-            try!(backup.store_one_source_entry(&try!(entry)));
+            let entry = entry?.clone();
```
Let's do the exclusions inside of `sources` rather than on the output, or at least make it a filter on the iter.

Actually, doing it in Sources will let us avoid traversing ignored directories.

sourcefrog commented on src/backup.rs:
```diff
@@ -26,22 +29,52 @@ struct Backup {
+                .clone()
+                .chars()
+                .skip(1)
+                .collect::<String>();  // remove leading '/‘ from path
```
This is a longwinded way to skip the first character - and, do we really want to?

sourcefrog commented on src/backup.rs:
```diff
@@ -26,22 +29,52 @@ struct Backup {
+
+            if excludes.is_some() && excludes.clone().unwrap().matches(&path).len() > 0 {
+                info!("Skipping {}", path);
+                continue;
```
Would be nice to count how many are skipped into the report.

SailReal commented:
@sourcefrog thanks for the feedback! Implemented all of the suggested changes...

sourcefrog commented on src/excludes.rs:
```diff
@@ -0,0 +1,70 @@
+                Ok(Some(builder.build()
+                    .chain_err(|| "Failed to build exclude patterns")?))
+            }
+            None => Ok(None)
```
Rather than an `Option<GlobSet>` can we make it return a `GlobSet` that matches nothing? That might make it easier for the callers.

sourcefrog commented on README.md:
```diff
@@ -47,6 +47,8 @@ source tree), the backup is considered *complete*.
 
     conserve backup /backup/home.cons ~
 
+Provide as option `--exclude` a list of [glob patterns](https://en.wikipedia.org/wiki/Glob_(programming) to exclude files or folders from backup.
```
Could you link to something giving the specific glob syntax supported by `GlobSet`?

sourcefrog replied:
We should also mention this in README.md.

sourcefrog replied:
Can you rephrase it to 

    Use `--exclude GLOB` to exclude files or directories from the backup.

sourcefrog commented on src/backup.rs:
```diff
@@ -26,17 +29,26 @@ struct Backup {
 
 impl BackupOptions {
+    pub fn with_excludes(excludes: Option<Vec<&str>>) -> Result<Self> {
+        let mut backup_options = BackupOptions::default();
```
This should consume a BackupOptions and return a new one, so the call looks like

    BackupOptions::default().with_excludes(...).backup(...)

sourcefrog commented on src/backup.rs:
```diff
@@ -26,17 +29,26 @@ struct Backup {
+
         for entry in source_iter {
-            try!(backup.store_one_source_entry(&try!(entry)));
+            try!(backup.store_one_source_entry(&entry?.clone()));
```
You shouldn't need to clone it here?

sourcefrog commented on src/bin/conserve.rs:
```diff
@@ -104,6 +106,12 @@ fn make_clap<'a, 'b>() -> clap::App<'a, 'b> {
+        .arg(Arg::with_name("exclude")
+            .long("exclude")
+            .short("e")
+            .global(true)
```
This actually shouldn't be global: it will apply to many commands but not all, for example not `init`.

sourcefrog commented on src/bin/conserve.rs:
```diff
@@ -104,6 +106,12 @@ fn make_clap<'a, 'b>() -> clap::App<'a, 'b> {
+            .short("e")
+            .global(true)
+            .takes_value(true)
+            .help("Exclude Files that matches the provided pattern [glob1,glob2,..]"))
```
Can you let this be supplied multiple times, and combine them?

sourcefrog replied:
Also "Files" doesn't need a capital.

sourcefrog commented on src/bin/conserve.rs:
```diff
@@ -180,16 +188,26 @@ fn init(subm: &ArgMatches, _report: &Report) -> Result<()> {
+                        report)
+        }
+        None => {
+            BackupOptions::default()
```
For example here, if you make `with_excludes` modify an existing `BackupOptions`, you won't need to repeat so much code.

sourcefrog commented on src/sources.rs:
```diff
@@ -92,8 +96,21 @@ impl Iter {
-            children.push((entry.file_name(), ft.is_dir()));
+
+            if self.excludes.is_some() &&
+                self.excludes.clone().unwrap().matches(&entry_name).len() > 0 {
```
I don't think you should need to clone it?

I think you can just use `is_match` rather than checking the len?

SailReal commented:
Thanks for reviewing again!
Implemented the improvements and also `--exclude` for `restore`, `ls`, and `list-sources`.

Added in [sources.rs](https://github.com/SailReal/conserve/blob/e66b9f3ed3cf45c5f35beef2ad28f73e0cad1cd5/src/sources.rs#L102) a //FIXME. 
Don't know why I'm using there the name and not the complete path to match against the pattern:
```rust
if self.excludes.is_match(&entry_name) { //FIXME @SailReal use path instead of filename
```

SailReal commented:
Thanks for reviewing again!
Implemented the improvements and also `--exclude` for `restore`, `ls`, and `list-source`.
Would you like to review again? 😎

sourcefrog commented:
Sorry for the delay, will look again soon

On Nov 29, 2017 4:02 PM, "Julian Raufelder" <notifications@github.com>
wrote:

> Thanks for reviewing again!
> Implemented the improvements and also --exclude for restore, ls, and
> list-source.
> Would you like to review again? 😎
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sourcefrog/conserve/pull/64#issuecomment-348038093>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAVI8_hsodbcIN8KaOpJgAeY1IzGRnM_ks5s7fCYgaJpZM4QVB8R>
> .
>

sourcefrog commented on src/index.rs:
```diff
@@ -120,7 +121,7 @@ impl IndexBuilder {
         let mut af = try!(AtomicFile::new(hunk_path));
         let compressed_len = report.measure_duration("index.compress",
-            || Snappy::compress_and_write(json_string.as_bytes(), &mut af))?;
+                                                     || Snappy::compress_and_write(json_string.as_bytes(), &mut af))?;
```
(nit) I'm not sure what tool you're using but that doesn't look like an improvement.

sourcefrog commented on src/index.rs:
```diff
@@ -245,7 +248,20 @@ impl Iter {
+                if self.excludes.is_match(&entry.apath) {
+                    match entry.kind {
+                        IndexKind::Dir => self.report.increment("skipped.excluded.directories", 1),
+                        IndexKind::Symlink => self.report.increment("skipped.excluded.directories", 1),
```
Should be `symlinks` not `directories` on this line.

sourcefrog commented on src/index.rs:
```diff
@@ -245,7 +248,20 @@ impl Iter {
+                }
+                return true;
+            })
+            .collect::<Vec<Entry>>().into_iter();
```
I don't understand why you collect it and then make an iter from that?

SailReal replied:
The result of `buffered_entries` has to be of type `vec::IntoIter<Entry>`. I have an `Vec<Entry>`and want to exclude some entries. Is there a better way to do that?

sourcefrog replied:
Oh OK. Perhaps we could later refactor `buffered_entries` but what you have here seems fine.

sourcefrog commented on src/restore.rs:
```diff
@@ -9,16 +9,34 @@ use std::path::Path;
+        Ok(RestoreOptions {
+            force_overwrite: self.force_overwrite,
+            band_id: self.band_id.clone(),
+            excludes: excludes::produce_excludes(exclude)?
```
Use `.. self` instead of repeating all the fields here.

sourcefrog commented on src/sources.rs:
```diff
@@ -89,19 +93,30 @@ impl Iter {
+                if ft.is_dir() {
+                    self.report.increment("skipped.excluded.directories", 1);
+                } else {
+                    self.report.increment("skipped.excluded.files", 1);
```
Here too you should handle symlinks.

sourcefrog commented on src/sources.rs:
```diff
@@ -89,19 +93,30 @@ impl Iter {
             let entry = try!(entry);
             let ft = try!(entry.file_type());
-            children.push((entry.file_name(), ft.is_dir()));
+            let mut path = dir_entry.apath.to_string().clone();
```
You shouldn't need the `.clone()`? Generally `.to_anything` returns a new object (https://aturon.github.io/style/naming.html) and it then shouldn't need to be cloned.

SailReal replied:
Can't get borrow the apath mutable without cloning because of the `to_string` method returns an reference: [to_string() in apath.rs](https://github.com/sourcefrog/conserve/blob/master/src/apath.rs#L33)

sourcefrog commented on src/sources.rs:
```diff
@@ -89,19 +93,30 @@ impl Iter {
+            // TODO: Don't be lossy, error if not convertible.
+            path.push_str(&entry.file_name().to_string_lossy());
+
+            if self.excludes.is_match(&path) {
```
I'm not sure but would it be reasonable to just match against the apath rather than constructing a separate path? How is it meant to be different?

SailReal replied:
This path is the apath for the entry. If I match the excludes against the apath of the entry then only the directories where excluded. The entry is a `DirType` and because of that it hasn't already an apath. So I just copied the apath creation for the children from the lines below (in origin line 100-114 [sources.rs#L100](https://github.com/sourcefrog/conserve/blob/master/src/sources.rs#L100)) to the actual location...

sourcefrog commented on src/sources.rs:
```diff
@@ -193,6 +206,7 @@ pub fn iter(source_dir: &Path, report: &Report) -> io::Result<Iter> {
         dir_deque: dir_deque,
         report: report.clone(),
         last_apath: None,
+        excludes: excludes.clone()
```
(minor) Probably reasonable to just pass `excludes` by value and consume it, rather than cloning.

sourcefrog commented on src/excludes.rs:
```diff
@@ -0,0 +1,68 @@
+    produce_globset(excludes)
+}
+
+pub fn produce_no_excludes() -> GlobSet {
```
Maybe `exclude_nothing`?

sourcefrog commented:
Something about the way the option is declared seems to make it consume all the following arguments, I think:

```
mbp@optimist ~/src/conserve➤ target/debug/conserve backup --exclude '/target/**' ~/Backup/conserve.c4/ .
error: The following required arguments were not provided:
    <archive>
    <source>

USAGE:
    conserve backup <archive> <source> --exclude <exclude>...

For more information try --help
```

It works if I put the option last.

SailReal commented:
Wow that strange behavior was because multiple was set to true. By removing multiples I could only pass one `--exclude` pattern but it doesn't matter where it was located (as expected)...

I looked into the github issues of clap and found some in this direction which was already closed...just updated the clap version and it worked...can you confirm that on your system?

sourcefrog commented:
> I looked into the github issues of clap and found some in this direction which was already closen...just updated the clap version and it worked...can you confirm that on your system?

Even at the tip of your branch and after `cargo update` I still seem to get an error when `--exclude` comes before the filenames...

sourcefrog commented:
`multiple` seems strangely defined in clap, and maybe not the right thing here: https://docs.rs/clap/2.29.0/clap/struct.Arg.html#method.multiple.

I think I see what needs to be done, I'll try adding a revision to this branch.
